// @hyperledger/fabric-gateway/dist/chaincodeeventsbuilder.js
"use strict";
/*
 * Copyright 2021 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _ChaincodeEventsBuilder_instances, _ChaincodeEventsBuilder_options, _ChaincodeEventsBuilder_newChaincodeEventsRequestProto, _ChaincodeEventsBuilder_getStartPosition;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChaincodeEventsBuilder = void 0;
const chaincodeeventsrequest_1 = require("./chaincodeeventsrequest");
const gateway_pb_1 = require("./protos/gateway/gateway_pb");
const ab_pb_1 = require("./protos/orderer/ab_pb");
class ChaincodeEventsBuilder {
    constructor(options) {
        _ChaincodeEventsBuilder_instances.add(this);
        _ChaincodeEventsBuilder_options.set(this, void 0);
        __classPrivateFieldSet(this, _ChaincodeEventsBuilder_options, options, "f");
    }
    build() {
        return new chaincodeeventsrequest_1.ChaincodeEventsRequestImpl({
            client: __classPrivateFieldGet(this, _ChaincodeEventsBuilder_options, "f").client,
            signingIdentity: __classPrivateFieldGet(this, _ChaincodeEventsBuilder_options, "f").signingIdentity,
            request: __classPrivateFieldGet(this, _ChaincodeEventsBuilder_instances, "m", _ChaincodeEventsBuilder_newChaincodeEventsRequestProto).call(this),
        });
    }
}
exports.ChaincodeEventsBuilder = ChaincodeEventsBuilder;
_ChaincodeEventsBuilder_options = new WeakMap(), _ChaincodeEventsBuilder_instances = new WeakSet(), _ChaincodeEventsBuilder_newChaincodeEventsRequestProto = function _ChaincodeEventsBuilder_newChaincodeEventsRequestProto() {
    const result = new gateway_pb_1.ChaincodeEventsRequest();
    result.setChannelId(__classPrivateFieldGet(this, _ChaincodeEventsBuilder_options, "f").channelName);
    result.setChaincodeId(__classPrivateFieldGet(this, _ChaincodeEventsBuilder_options, "f").chaincodeName);
    result.setIdentity(__classPrivateFieldGet(this, _ChaincodeEventsBuilder_options, "f").signingIdentity.getCreator());
    result.setStartPosition(__classPrivateFieldGet(this, _ChaincodeEventsBuilder_instances, "m", _ChaincodeEventsBuilder_getStartPosition).call(this));
    return result;
}, _ChaincodeEventsBuilder_getStartPosition = function _ChaincodeEventsBuilder_getStartPosition() {
    const result = new ab_pb_1.SeekPosition();
    const startBlock = __classPrivateFieldGet(this, _ChaincodeEventsBuilder_options, "f").startBlock;
    if (startBlock != undefined) {
        const specified = new ab_pb_1.SeekSpecified();
        specified.setNumber(Number(startBlock));
        result.setSpecified(specified);
        return result;
    }
    result.setNextCommit(new ab_pb_1.SeekNextCommit());
    return result;
};
