// @hyperledger/fabric-gateway/dist/transactionparser.js
"use strict";
/*
 * Copyright 2021 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseTransactionEnvelope = void 0;
const util_1 = require("util");
const gateway_1 = require("./gateway");
const common_pb_1 = require("./protos/common/common_pb");
const proposal_pb_1 = require("./protos/peer/proposal_pb");
const proposal_response_pb_1 = require("./protos/peer/proposal_response_pb");
const transaction_pb_1 = require("./protos/peer/transaction_pb");
function parseTransactionEnvelope(envelope) {
    const payload = common_pb_1.Payload.deserializeBinary(envelope.getPayload_asU8());
    const header = (0, gateway_1.assertDefined)(payload.getHeader(), 'Missing header');
    return {
        channelName: parseChannelNameFromHeader(header),
        result: parseResultFromPayload(payload),
    };
}
exports.parseTransactionEnvelope = parseTransactionEnvelope;
function parseChannelNameFromHeader(header) {
    const channelHeader = common_pb_1.ChannelHeader.deserializeBinary(header.getChannelHeader_asU8());
    return channelHeader.getChannelId();
}
function parseResultFromPayload(payload) {
    const transaction = transaction_pb_1.Transaction.deserializeBinary(payload.getData_asU8());
    const errors = [];
    for (const transactionAction of transaction.getActionsList()) {
        try {
            return parseResultFromTransactionAction(transactionAction);
        }
        catch (err) {
            errors.push(err);
        }
    }
    throw Object.assign(new Error(`No proposal response found: ${(0, util_1.inspect)(errors)}`), {
        suppressed: errors,
    });
}
function parseResultFromTransactionAction(transactionAction) {
    const actionPayload = transaction_pb_1.ChaincodeActionPayload.deserializeBinary(transactionAction.getPayload_asU8());
    const endorsedAction = (0, gateway_1.assertDefined)(actionPayload.getAction(), 'Missing endorsed action');
    const responsePayload = proposal_response_pb_1.ProposalResponsePayload.deserializeBinary(endorsedAction.getProposalResponsePayload_asU8());
    const chaincodeAction = proposal_pb_1.ChaincodeAction.deserializeBinary(responsePayload.getExtension_asU8());
    const chaincodeResponse = (0, gateway_1.assertDefined)(chaincodeAction.getResponse(), 'Missing chaincode response');
    return chaincodeResponse.getPayload_asU8();
}
